File: java/src/test/java/com/example/hexagonal/adapter/inbound/ProductControllerTest.java
package com.example.hexagonal.adapter.inbound;

import static com.example.hexagonal.application.domain.ModelFixture.DEFAULT_ID;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.example.hexagonal.application.domain.ModelFixture;
import com.example.hexagonal.application.port.dto.DtoFixture;
import com.example.hexagonal.application.port.dto.DtoMapper;
import com.example.hexagonal.application.port.dto.ProductDto;
import com.example.hexagonal.application.port.inbound.ProductPort;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

@SpringBootTest
class ProductControllerTest {

  static ObjectMapper mapper = new ObjectMapper(); // Don't do in production; Declare a bean

  MockMvc mvc;

  @MockBean ProductPort mockUsecase;
  @Autowired DtoMapper dtoMapper;

  @Test
  void getProduct() throws Exception {
    Mockito
        .when(mockUsecase.getProduct(anyLong()))
        .thenReturn(ModelFixture.aProduct());

    mvc
        .perform(get("/api/products/{productId}", DEFAULT_ID)
            .accept("application/json")
        )
        .andDo(print())
        .andExpect(status().is2xxSuccessful());
  }

  @Test
  void updateProduct() throws Exception {
    final ProductDto dto = DtoFixture.aProductDtoOf(DEFAULT_ID);

    Mockito
        .when(mockUsecase.updateProduct(DEFAULT_ID, dto))
        .thenReturn(ModelFixture.aProduct());

    final byte[] content = mapper.writeValueAsBytes(dto);

    mvc
        .perform(put("/api/products/{productId}", DEFAULT_ID)
            .contentType("application/json")
            .accept("application/json")
            .content(content)
        )
        .andDo(print())
        .andExpect(status().is2xxSuccessful());
  }

  @BeforeEach
  void setup() {
    this.mvc = MockMvcBuilders
        .standaloneSetup(new ProductController(mockUsecase, dtoMapper))
        .build();
  }
}

File: kotlin/src/test/kotlin/com/example/hexagonal/adapter/inbound/ProductControllerTest.kt
package com.example.hexagonal.adapter.inbound

import com.example.hexagonal.TestConfig
import com.example.hexagonal.application.domain.Product
import com.example.hexagonal.application.port.dto.DtoMapper
import com.example.hexagonal.application.port.inbound.ProductPort
import com.fasterxml.jackson.databind.ObjectMapper
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.extensions.spring.SpringExtension
import io.mockk.every
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.context.annotation.Import
import org.springframework.test.context.TestConstructor
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultHandlers
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.springframework.test.web.servlet.setup.MockMvcBuilders

@WebMvcTest(ProductController::class)
@Import(TestConfig::class, DtoMapper::class)
@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
internal open class ProductControllerTest(
    val mockUsecase: ProductPort,
    val dtoMapper: DtoMapper,
) : DescribeSpec() {
    override fun extensions() = listOf(SpringExtension)

    init {
        describe("Product API") {
            context("calls 'getProduct' API") {
                it("should success with 2xx status") {
                    every { mockUsecase.getProduct(any()) } returns Product.aProduct()

                    mvc
                        .perform {
                            MockMvcRequestBuilders
                                .get("/api/products/{productId}", fixture.id)
                                .accept("application/json")
                                .buildRequest(it)
                        }
                        .andDo(MockMvcResultHandlers.print())
                        .andExpect(MockMvcResultMatchers.status().is2xxSuccessful())
                }
            }

            context("calls 'updateProduct' API") {
                it("should success with 2xx status") {
                    val dto = Product.aProduct()
                    every { mockUsecase.updateProduct(any(), any()) } returns dto

                    mvc
                        .perform {
                            MockMvcRequestBuilders
                                .put("/api/products/{productId}", fixture.id)
                                .contentType("application/json")
                                .accept("application/json")
                                .content(mapper.writeValueAsBytes(dto))
                                .buildRequest(it)
                        }
                        .andDo(MockMvcResultHandlers.print())
                        .andExpect(MockMvcResultMatchers.status().is2xxSuccessful())
                }
            }
        }
    }

    val mvc: MockMvc = MockMvcBuilders
        .standaloneSetup(ProductController(mockUsecase, dtoMapper))
        .build()

    val fixture: Product = Product.aProduct()

    companion object {
        val mapper = ObjectMapper()
    }
}
